wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs

```diff
@@ -1,43 +1,233 @@
-        { pwlGetWalletIds  = askWalletSnapshot >>= \ws -> migrate $ getWalletAddresses ws
+        { _pwlCreateWallet  = pwlCreateWallet
+        , _pwlGetWalletIds  = pwlGetWalletIds
+        , _pwlGetWallet     = pwlGetWallet
+        , _pwlUpdateWallet  = pwlUpdateWallet
+        , _pwlDeleteWallet  = pwlDeleteWallet
+
+        , _pwlCreateAccount = pwlCreateAccount
+        , _pwlGetAccounts   = pwlGetAccounts
+        , _pwlGetAccount    = pwlGetAccount
+        , _pwlUpdateAccount = pwlUpdateAccount
+        , _pwlDeleteAccount = pwlDeleteAccount
+
+        , _pwlGetAddresses  = pwlGetAddresses
         }
 
+
 -- | Initialize the active wallet.
 -- The active wallet is allowed all.
 bracketActiveWallet
-    :: forall ctx m n a. (MonadMask m, WalletDbReader ctx n, MonadIO n, MonadThrow n)
-    => PassiveWalletLayer n
+    :: forall ctx m n a. (MonadMask n, MonadLegacyWallet ctx m)
+    => PassiveWalletLayer m
     -> WalletDiffusion
-    -> (ActiveWalletLayer n -> m a) -> m a
+    -> (ActiveWalletLayer m -> n a) -> n a
 bracketActiveWallet walletPassiveLayer walletDiffusion =
     bracket
       (return ActiveWalletLayer{..})
       (\_ -> return ())
 
+------------------------------------------------------------
+-- Wallet
+------------------------------------------------------------
+
+pwlCreateWallet
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => NewWallet
+    -> m Wallet
+pwlCreateWallet NewWallet{..} = do
+
+    let spendingPassword = fromMaybe mempty $ coerce newwalSpendingPassword
+    let backupPhrase     = coerce newwalBackupPhrase
+
+    initMeta    <- CWalletMeta  <$> pure newwalName
+                                <*> migrate newwalAssuranceLevel
+                                <*> pure 0
+
+    let walletInit = CWalletInit initMeta backupPhrase
+
+    wallet      <- newWalletHandler newwalOperation spendingPassword walletInit
+    wId         <- migrate $ cwId wallet
+
+    -- Get wallet or throw if missing.
+    maybeThrow (MissingWallet wId) =<< pwlGetWallet wId
+  where
+    -- | We have two functions which are very similar.
+    newWalletHandler :: WalletOperation -> PassPhrase -> CWalletInit -> m CWallet
+    newWalletHandler CreateWallet  = newWallet
+    newWalletHandler RestoreWallet = restoreWallet
+
+
+pwlGetWalletIds
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => m [WalletId]
+pwlGetWalletIds = do
+    ws          <- askWalletSnapshot
+    migrate $ getWalletAddresses ws
+
+pwlGetWallet
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => WalletId
+    -> m (Maybe Wallet)
+pwlGetWallet wId = do
+    ws          <- askWalletSnapshot
+
+    cWId        <- migrate wId
+    wallet      <- V0.getWallet cWId
+
+    pure $ do
+        walletInfo  <- getWalletInfo cWId ws
+        migrate (wallet, walletInfo)
+
+pwlUpdateWallet
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => WalletId
+    -> WalletUpdate
+    -> m Wallet
+pwlUpdateWallet wId wUpdate = do
+    walletDB    <- askWalletDB
```

Comment:
Is there any document where we elaborate a bit more on the usage of the db vs a snapshot of it? I'd like to understand the rational behind this :thinking:

Reply:
I agree. It's a little confusing. I recall @adinapoli-iohk saying that we acquire a snapshot of the DB, so that a single request returns a consistent answer. It's sort of like a read-lock on a SQL database table for a transaction.

Of course, we must get a *new* snapshot if we perform any updates, otherwise we run the risk of stale reads!

Reply:
There isn't a document AFAIK, but since we are going to change this anyway, not sure if there is anything better to write then checking out https://github.com/input-output-hk/cardano-sl/blob/develop/wallet/src/Pos/Wallet/Web/State/State.hs
If that is confusing as well, I can write a couple of lines in some documentation, but essentially we use the snapshot for reading and the db for writing.